test/framework/kubernetesversions/template.go

```diff
@@ -116,91 +134,97 @@ func GenerateCIArtifactsInjectedTemplateForDebian(input GenerateCIArtifactsInjec
 	return kustomizedTemplate, nil
 }
 
-func generateKustomizeVersionsYaml(kcpName, kubeadmTemplateName, kubeadmConfigName string) ([]byte, error) {
-	kcp := generateKubeadmControlPlane(kcpName)
-	kubeadm := generateKubeadmConfigTemplate(kubeadmTemplateName)
-	kcpYaml, err := yaml.Marshal(kcp)
-	if err != nil {
+func generateKustomizationYAML(input GenerateCIArtifactsInjectedTemplateForDebianInput) ([]byte, error) {
+	var kustomizationYamlBytes bytes.Buffer
+	if err := kustomizationTemplate.Execute(&kustomizationYamlBytes, input); err != nil {
 		return nil, err
 	}
-	kubeadmYaml, err := yaml.Marshal(kubeadm)
-	if err != nil {
-		return nil, err
-	}
-	fileStr := string(kcpYaml) + yamlSeparator + string(kubeadmYaml)
-	if kubeadmConfigName == "" {
-		return []byte(fileStr), nil
-	}
-
-	kubeadmConfig := generateKubeadmConfig(kubeadmConfigName)
-	kubeadmConfigYaml, err := yaml.Marshal(kubeadmConfig)
-	if err != nil {
-		return nil, err
-	}
-	fileStr = fileStr + yamlSeparator + string(kubeadmConfigYaml)
-
-	return []byte(fileStr), nil
+	return kustomizationYamlBytes.Bytes(), nil
 }
 
-func generateKubeadmConfigTemplate(name string) *cabpkv1.KubeadmConfigTemplate {
-	kubeadmSpec := generateKubeadmConfigSpec()
-	return &cabpkv1.KubeadmConfigTemplate{
-		TypeMeta: metav1.TypeMeta{
-			Kind:       "KubeadmConfigTemplate",
-			APIVersion: cabpkv1.GroupVersion.String(),
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Name: name,
-		},
-		Spec: cabpkv1.KubeadmConfigTemplateSpec{
-			Template: cabpkv1.KubeadmConfigTemplateResource{
-				Spec: *kubeadmSpec,
-			},
-		},
-	}
+type jsonPatch struct {
+	Op    string      `json:"op"`
+	Path  string      `json:"path"`
+	Value interface{} `json:"value"`
 }
 
-func generateKubeadmConfig(name string) *cabpkv1.KubeadmConfig {
-	kubeadmSpec := generateKubeadmConfigSpec()
-	return &cabpkv1.KubeadmConfig{
-		TypeMeta: metav1.TypeMeta{
-			Kind:       "KubeadmConfig",
-			APIVersion: kcpv1.GroupVersion.String(),
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Name: name,
-		},
-		Spec: *kubeadmSpec,
+// generateInjectScriptJSONPatch generates a JSON patch which injects a script
+// * objectKind: is the kind of the object we want to inject the script into
+// * objectName: is the name of the object we want to inject the script into
+// * jsonPatchPathPrefix: is the prefix of the 'files' and `preKubeadmCommands` arrays where we append the script
+// * scriptPath: is the path where the script will be stored at
+// * scriptContent: content of the script
+func generateInjectScriptJSONPatch(sourceTemplate []byte, objectKind, objectName, jsonPatchPathPrefix, scriptPath, scriptContent string) ([]byte, error) {
+	filesPathExists, preKubeadmCommandsPathExists, err := checkIfArraysAlreadyExist(sourceTemplate, objectKind, objectName, jsonPatchPathPrefix)
```

Comment:
This is where it gets pretty hacky. I couldn't find a way via JSON patch to just "create an array if it isn't there". That's why I have to check first if the arrays already exist and only add them if they don't exist before.

I could have always set the arrays to an empty array but this would have overwritten every preexisting array elements.


I thought about just parsing the relevant objects, changing them and generating a JSON patch for the diff, maybe that would be better.

What do you think?